src/shell/shell_misc.cpp

```diff
@@ -389,106 +390,36 @@ void DOS_Shell::InputCommand(char * line) {
-				if (w >= wf) goto overflow;
-				*w++ = '%';
-				if (*r != 0) r++;
-				else break;
-			}
-			else {
-				const char* name = r; /* store pointer, 'r' is first char of the name following '%' */
-				int spaces = 0,chars = 0;
-				if (isalpha(*r) || *r == ' ') { /* must start with a letter. space is apparently valid too. (Win95) */
-					if (*r == ' ') spaces++;
-					else if (isalpha(*r)) chars++;
-					r++;
-					while (*r != 0 && *r != '%') {
-						if (*r == ' ') spaces++;
-						else chars++;
-						r++;
-					}
-				}
-				if (*r == '%' && ((spaces > 0 && chars == 0) || (spaces == 0 && chars > 0))) {
-					std::string temp2;
-					/* valid name found. substitute */
-					*r++ = 0; /* ASCIIZ snip */
-					if (GetEnvStr(name,temp2)) {
-						size_t equ_pos = temp2.find_first_of('=');
-						if (equ_pos != std::string::npos) {
-							const char *base = temp2.c_str();
-							const char *value = base + equ_pos + 1;
-							const char *fence = base + temp2.length();
-							assert(value >= base && value <= fence);
-							size_t len = (size_t)(fence-value);
-							if ((w+len) > wf) goto overflow;
-							memcpy(w,value,len);
-							w += len;
-						}
-					}
-				}
-				else {
-					/* nope. didn't find a valid name */
-					while (*r == ' ') r++; /* skip spaces */
-					name--; /* step "name" back to cover the first '%' we found */
-					for (const char* c = name; c < r;) {
-						if (w >= wf) goto overflow;
-						*w++ = *c++;
-					}
-				}
-			}
+/* Note: Buffer pointed to by "line" must be at least CMD_MAXLINE+1 bytes long! */
+void DOS_Shell::ProcessCmdLineEnvVarStitution(char* line) {
+	std::string text = std::string(line);
+	static std::regex env("\\%([^%0-9][^%]*)?%");
+	std::smatch match;
+	while (std::regex_search(text, match, env)) {
+		std::string temp1, temp2;
+		temp1 = match[1].str();
+		if (!temp1.size()) {
+			/* replace %% with ASCII 8, then (eventually) % */
+			text.replace(match[0].first, match[0].second, std::string(1, 8));
+			continue;
 		}
-		else {
-			if (w >= wf) goto overflow;
-			*w++ = *r++;
+		trim(temp1);
+		if (temp1.size() && GetEnvStr(temp1.c_str(), temp2)) {
+			size_t equ_pos = temp2.find_first_of('=');
+			if (equ_pos != std::string::npos) {
+				const char* base = temp2.c_str();
+				const char* value = base + equ_pos + 1;
+				text.replace(match[0].first, match[0].second, value);
+			}
+			else
+				text.replace(match[0].first, match[0].second, "");
 		}
+		else
+			text.replace(match[0].first, match[0].second, "");
 	}
-	/* complete the C-string */
-	assert(w <= wf);
-	*w = 0;
-	/* copy the string back over the buffer pointed to by line */
-	{
-		size_t out_len = (size_t)(w+1-temp); /* length counting the NUL too */
-		assert(out_len <= CMD_MAXLINE);
-		memcpy(line,temp,out_len);
-	}
-	/* success */
-	return;
-overflow:
-	*line = 0; /* clear string (C-string chop with NUL) */
-	WriteOut("Command input error: string expansion overflow\n");
+	std::replace(text.begin(), text.end(), (char)8, '%');
```

Comment:
Very interesting trick! I suppose char-8 is an an illegal / impossible-to-type character that will not occur on the command prompt.

Reply:
Further up when you first substitute 8 in-place, perhaps we could turn 8 into a variable:

 `constexpr char surrogate_percent = 8`

Give it's used as a stand-in for the percent symbol?  This will help readability where we have the `(1, 8)`..where (unless one has these API's memorized) might not know if it's a range.. a length.. or a char value.

Reply:
Sure. I put `constexpr char surrogate_percent = 8` as you suggested.